### ISSUE TITLE
Use title for oneOf union strategy

### ISSUE THREAD

**Issue by connorjs** [awslabs/smithy#483]
TL;DR - Can we use `title` for the inlined object generated as part of the `oneOf` union strategy?

---

_Disclaimer: I am opening my first issue for Smithy._

I recently started using Smithy and primarily generate Open API specifications. For context, I use [Redocly](https://github.com/Redocly/redoc) for documentation generation. When using Smithy's union type, I noticed that the generated Open API spec using inline types within the `oneOf` attribute (default strategy). This results in the same **display issue** as mentioned in [this Redocly/redoc issue](https://github.com/Redocly/redoc/issues/932), but is unrelated w.r.t. problem/solutions (I think). It showcases my desired outcome (display) though.

Glancing through the source code, I arrived at `JsonSchemaShapeVisitor`. I believe we would [add title in this code block](https://github.com/awslabs/smithy/blob/1851a85f3d503af87576280e8ad086ee1eca4889/smithy-jsonschema/src/main/java/software/amazon/smithy/jsonschema/JsonSchemaShapeVisitor.java#L192-L202). Due to the simple fix (as I see it), I can contribute the change. However, I want to confirm this solution is "correct."

1. I understand `title` to be descriptive in nature, but are there other consequences/implications of this attribute? I could not understand more from the [Swagger documentation](https://swagger.io/specification/#schemaObject) though.
2. I propose to add the member name as the title. This guarantee uniqueness (within the `oneOf`) and makes the most sense when reading: "One of foo, bar, or baz" for the following union.
    ```
    union MyUnion {
        foo: String,
        bar: String,
        baz: String,
    } 
    ```
3. I have not looked at tests yet, but I expect I would add to an existing case or create a title-specific one.
4. I believe this change should be backwards compatible given the nature of title (related to my first question in this list), but I would like confirmation.

Thanks in advance!

**Comment by mtdowling** [comment]
So the benefit of adding the `title` property is for better doc generation, correct? I don't see how this would be a breaking change, so I'm open to it, and contributions are always welcome!

> This guarantee uniqueness (within the oneOf) 

How so? As far as I can tell, it's just a documentation kind of property. Are you just saying that adding `title` will inherently have this effect in Redoc documentation?

**Comment by connorjs** [comment]
> So the benefit of adding the `title` property is for better doc generation, correct? I don't see how this would be a breaking change, so I'm open to it, and contributions are always welcome!

We are on the same page then 😀. I'll get started on it this week then.

> This guarantee uniqueness (within the oneOf)

To word it better: I propose using the member name as the title. Selecting the member name guarantees that each title is unique within a oneOf because each member is unique.

I included this comment mostly as rationale/extra defense for why I propose using the member name for the title. A trait could also be used, but I would prefer this to happen automatically as it affects Open API documentation only per my understanding.

Thanks!

**Comment by llgerard** [comment]
@connorjs @mtdowling when looking into this decision, did you explore the possibility to use a discriminator/tag? I am trying to find if my proposal makes sense: https://github.com/awslabs/smithy/issues/1082

**Comment by connorjs** [comment]
I believe this issue focused on a specific use case for generating documentation from an Open API spec originally written in Smithy. Therefore, I think the discriminator issue, which involves actual modeling, stands apart from this documentation-focused issue.
